IA Reader Navigation: adding subfiltering

develric opened the pull request:
This PR is one of a set of PRs which aim to supersede the #10716 decomposing it in more manageable/testable/reviewable PRs (cc @osullivanchris ).

The current one is based and requires the #10783 

Visually we have added a subfiltering component that should be available only in the `FOLLOWING` tab. Below you can see the subfiltering component

<p align="center">
  <img width="300" src=https://user-images.githubusercontent.com/47797566/68861160-413fc180-06eb-11ea-86fc-b796db7eb87b.png>
</p>

Below what you get tapping on each of the above

| Tag/Site selector  | Settings selector |
|---|---|
|![image](https://user-images.githubusercontent.com/47797566/68861257-7ea44f00-06eb-11ea-99a4-4b2fc0e484b6.png)|![image](https://user-images.githubusercontent.com/47797566/68861308-97146980-06eb-11ea-822e-2261c4cee14a.png)|

## Notes
- Feature flag is again active only in zalpha flavor. Important goal is to make the feature flag have no impact in code base when disabled (vanilla flavor)! So I appreciate a special eye on the code in this direction :)
- The settings screen is not changed, only the botton from which to access to it was moved to the subfilter bar
- Similarly to the bottom sheet for the create post/pages in my site, also this one is pretty basic. There is some design work ongoing in the direction of creating a standard component we can use for the Bottom Sheet paradigm
- Could not find a better way to filter out the Discover from the followed sites in the bottom sheet than filter on the URL (better suggestions accepted :) )
- Left for now (tracking it in a separate issue) a different behavior that was already present but less impacting with the previous IA. When adding a new tag in settings and coming back to the reader, the new added tag is selected and posts shown. This does not happen when adding a followed site from settings
- Release notes will be updated when releasing in develop removing the feature flag

## To test
### With the vanilla flavor
- Use the Vanilla flavor and smoke test to check the original behavior didn't changed (doing same steps with same user account in parallel on emulator/device with last release and this one would be great)
- Check the behavior of My Site > Comments and My Site > People did not change
- Check the behavior of Reader nested navigation did not change

### With the zalpha flavor
- Check that you have only the FOLLOWING, DISCOVER, LIKES and SAVED tabs in the reader (NOTE: you will also get AUTOMATTIC tab as last tab in case you login with an Automattic account)
- Check that the subfilter bar is only available when you are in the FOLLOWING tab
- Try to open the settings and try to add/remove a TAG , when coming back in the reader you should get that one selected or the Following selected if you removed the tag yo were looking at
- Try to open the settings and try to follow/unfollow a site, when coming back to the reader you should get the same page where you were before
- Select a site from Discover tab, you should get a list of post from that site in a new activity. Come back from it and you should get the reader stay in the same tab you left before
- smoke test the search icon
- Smoke test the reader in general (move between tabs, use the subfilter to select different sources, open posts from the reader, follow/unfollow tags/sites etc...)

PR submission checklist:

- [x] I have considered adding unit tests where possible.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid commented:
thanks for the changes @develric! I started with a quick test of the changes before looking at the code and I found a bug.
Steps to reproduce:
* Go to Reader/Discover
* Go to "My Site"
* Go back to the Reader
* The sub-toolbar is visible but the items are not clickable

![bug_in_top_toolbar](https://user-images.githubusercontent.com/1079756/68871189-5b35d000-06fc-11ea-9602-1a032d01215a.gif)

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/reader/subfilter/viewholders/SubfilterListItemViewHolder.kt:
```diff
@@ -0,0 +1,35 @@
+        itemText?.let {
+            if (filter.isSelected) {
+                @Suppress("DEPRECATION")
+                it.setTextColor(parent.resources.getColor(R.color.primary))
```
I think you should use `.setTextColor(ContextCompat.getColor(it, R.color.primary))` instead so you don't have to suppress the warning.

develric replied:
Better, thanks for pointing out! 😊Changed here fed0e186b3eb4f9be44740543af84242f01b2cb3

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostListViewModel.kt:
```diff
@@ -80,6 +108,104 @@ class ReaderPostListViewModel @Inject constructor(
+        }
+    }
+
+    private fun onSubfilterClicked(filter: SubfilterListItem) {
```
I don't think the `when` is necessary since the other types are not clickable, right?

develric replied:
I was guided by an attempt to be more explicit, but I agree simpler is better for now. Changed here fed0e186b3eb4f9be44740543af84242f01b2cb3

planarvoid replied:
👍 If we ever have more than 1 click action, I'd rather have more click handle methods

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -378,9 +393,92 @@ public void onCreate(Bundle savedInstanceState) {
         mViewModel.start(mCurrentTag);
     }
 
+    private void setCurrentSubfilter(SubfilterListItem subfilterListItem, boolean requestNewerPosts) {
```
Would it be possible to move this logic into the view model and only expose an endpoint with `LiveData<CurrentTagInfo>` and keep the object as a field in here?

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -768,9 +886,52 @@ public void onClick(View view) {
+        ReaderActivityLauncher.showReaderSubs(getActivity());
+    }
+
+    private boolean isFollowing() {
```
Could we maybe pull this logic somewhere else? We can pass `mCurrentTag` and a nullable `currentFilter` as parameters. Maybe to ReaderUtils? Or something injected?

develric replied:
done here cb0b1590629e59f73babc5f04b9522100bf4d687

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -1686,6 +1885,20 @@ private boolean hasCurrentTag() {
+        setCurrentTag(tag, false);
+    }
+
+    private boolean isValidTagForSharedPrefs(ReaderTag tag) {
```
I'd also extract this one :-)

develric replied:
yep. done here cb0b1590629e59f73babc5f04b9522100bf4d687

develric commented:
Hi @planarvoid , I should have followed up to your comments and fixed (hopefully) the issue you found (thanks for that 🙇‍♂️). Let me know wdyt 😊.

planarvoid commented:
Hey, thanks for the changes! Overall it looks and works pretty well 👍 
Testing results:
With the vanilla flavor
> Use the Vanilla flavor and smoke test to check the original behavior didn't changed (doing same steps with same user account in parallel on emulator/device with last release and this one would be great)
- As discussed, the order of the items in the spinner has changed, it should be `Automattic, My likes, Discover and Followed`. Now it's `Automattic, Discover, Followed and My Likes`
- When I turn off the FF and reinstall the app, the items in the spinner are wrong (they "likes" instead of "My Likes" and so on), this is an issue with the database we've discussed before.
> Check the behavior of My Site > Comments and My Site > People did not change

👍 
> Check the behavior of Reader nested navigation did not change

👍 
> With the zalpha flavor
> Check that you have only the FOLLOWING, DISCOVER, LIKES and SAVED tabs in the reader

I also have the `AUTOMATTIC` tab there, I guess that's correct? 
> Check that the subfilter bar is only available when you are in the FOLLOWING tab

👍 
> Try to open the settings and try to add/remove a TAG , when coming back in the reader you should get that one selected or the Following selected if you removed the tag yo were looking at

👍 
> Try to open the settings and try to follow/unfollow a site, when coming back to the reader you should get the same page where you were before

This works, however when I unfollow a site that I have in the subfilter, it stays there and I don't think it makes sense. It disappears from the list of sites in the subfilter but it's still the selected site.
> Select a site from Discover tab, you should get a list of post from that site in a new activity. Come back from it and you should get the reader stay in the same tab you left before

👍 
> smoke test the search icon

👍 
> Smoke test the reader in general (move between tabs, use the subfilter to select different sources, open posts from the reader, follow/unfollow tags/sites etc...)

When I rotate my phone and click on the `filter`, I get this view. I think we should maybe scroll it to cover most of the screen?
![Screenshot_1574757264](https://user-images.githubusercontent.com/1079756/69612444-f64d7480-102f-11ea-92aa-31b1b162b1c5.png)

develric commented:
Hi @planarvoid, 
I made some changes as per below comments; as discussed on Slack I would prefer to keep the last two points as open items to work on later and have this PR progress toward feature branch and then develop so we can ship it to internal alpha testing and start getting feedbacks without grow the feature branch more. Let me know your thoughts and thanks a lot for all the much appreciated support in this task! 🙇‍♂️ 

> As discussed, the order of the items in the spinner has changed, it should be Automattic, My likes, Discover and Followed. Now it's Automattic, Discover, Followed and My Likes

Actually in my device with the store app version I get Automattic, Discover, Followed and My Likes. You mentioned in slack that you were testing the app also in different languages, I think the order used in the original version is alphabetical, could it be that the different order is due to a difference in language used? Otherwise I'm a bit out of clues and need to check more 😅.

> When I turn off the FF and reinstall the app, the items in the spinner are wrong (they "likes" instead of "My Likes" and so on), this is an issue with the database we've discussed before.

I was addressing already the reverse direction (from FF off to FF on) but I agree we need to check for both! I changed to try to detect the "downgrade" based on Saved tag display name (not great but should work)

> I also have the AUTOMATTIC tab there, I guess that's correct?

Yep you are right and this is expected. I changed the PR description to reflect this, thanks 😊

> This works, however when I unfollow a site that I have in the subfilter, it stays there and I don't think it makes sense. It disappears from the list of sites in the subfilter but it's still the selected site.

I already noticed this and traced in the PR description notes as open item. I will create a GH issue for this to work on later.

> When I rotate my phone and click on the filter, I get this view. I think we should maybe scroll it to cover most of the screen?

Also for this one, since there is an ongoing design to create a bottom sheet component I would not spend time now but work directly on the new component.

One additional note from my side:

While fixing the above I noticed a bug 🤯; steps to reproduce (and check it is no more present):
- select a tag from the subfilter (bottom part of the list)
- rotate the device
- the subfilter does not update and you get no label in it

I should have fixed it now but needed to change the isFollowing function that is quite important one for all the logic, so would be good to make another full round check, sorry for this!

planarvoid commented:
> I think the order used in the original version is alphabetical

I've double checked that and you're right 🤦‍♂, even the tag is under the `Saved` posts

> I already noticed this and traced in the PR description notes as open item. I will create a GH issue for this to work on later.

Sounds good 👍 

> Also for this one, since there is an ongoing design to create a bottom sheet component I would not spend time now but work directly on the new component.

Makes sense 👍

planarvoid commented:
thanks for the changes @develric , while testing your newly discovered bug, I've actually discovered another one :-)
Steps to reproduce: 
* Have a tag selected
* Click on the filter
* Rotate your device
* Try clicking on any site/tag and nothing happens
* You can't leave the screen unless you press on the back button

One more thing I find a little weird is that the bottom buttons disappear when the bottom sheet is opened but that might be an emulator issue.

develric commented:
Hi @planarvoid, Good catch! I was able to reproduce (thanks for documenting the steps). As discussed if you have no more points for me to address we can merge this PR in the feature branch and fix the bug in a dedicated PR.
